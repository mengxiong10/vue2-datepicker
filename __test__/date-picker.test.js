import { shallowMount, createWrapper, mount } from '@vue/test-utils';
import { format, parse } from 'date-fns';
import Popup from '../src/popup';
import DatePicker from '../src/date-picker';

let wrapper;

afterEach(() => {
  wrapper.destroy();
});

describe('DatePicker', () => {
  it('feat: open and close the popup', async () => {
    wrapper = mount(DatePicker, {
      attachTo: document.body,
    });
    expect(wrapper.find('.mx-datepicker-popup').exists()).toBe(false);
    // expect click input should show the popup
    const input = wrapper.find('input');
    await input.trigger('click');
    expect(wrapper.find('.mx-datepicker-popup').exists()).toBe(true);
    // expect click out side should hide the popup
    const bodyWrapper = createWrapper(document.body);
    await bodyWrapper.trigger('mousedown');
    expect(wrapper.find('.mx-datepicker-popup').exists()).toBe(false);
    // expect keydown down should show the popop
    await input.trigger('focus');
    await input.trigger('keydown.down');
    expect(wrapper.find('.mx-datepicker-popup').exists()).toBe(true);
    // expoce keydown tab should hide the popup
    await input.trigger('keydown.tab');
    expect(wrapper.find('.mx-datepicker-popup').exists()).toBe(false);
  });

  it('feat: should init panel and calendar when reopen', async () => {
    wrapper = mount(DatePicker, {
      propsData: {
        defaultValue: new Date(2019, 9, 1),
        open: true,
      },
    });
    const yearBtn = wrapper.find('.mx-btn-current-year');
    await yearBtn.trigger('click');
    // change to year panel
    expect(wrapper.find('.mx-calendar-panel-year').exists()).toBe(true);
    await wrapper.setProps({ open: false });
    await wrapper.setProps({ open: true });
    expect(wrapper.find('.mx-calendar-panel-year').exists()).toBe(false);
  });

  it('prop: open', async () => {
    wrapper = mount(DatePicker, {
      propsData: {
        open: false,
      },
    });
    const { vm } = wrapper;
    vm.openPopup();
    expect(wrapper.find('.mx-datepicker-popup').exists()).toBe(false);
    await wrapper.setProps({ open: true });
    expect(wrapper.find('.mx-datepicker-popup').exists()).toBe(true);
    vm.closePopup();
    expect(wrapper.find('.mx-datepicker-popup').exists()).toBe(true);
    const emitted = wrapper.emitted();
    expect(emitted['update:open'][0][0]).toBe(true);
    expect(emitted['update:open'][1][0]).toBe(false);
  });

  it('prop: disabled(should not show the popup)', () => {
    wrapper = mount(DatePicker, {
      propsData: {
        disabled: true,
      },
    });
    const input = wrapper.find('input');
    expect(input.attributes('disabled')).toBe('disabled');
    input.trigger('click');
    expect(wrapper.find('.mx-datepicker-popup').exists()).toBe(false);
  });

  it('prop: clearable', () => {
    wrapper = shallowMount(DatePicker, {
      propsData: {
        value: new Date(2019, 4, 10),
        defaultValue: new Date(2019, 4, 10),
        clearable: false,
      },
      // calendar-panel: default-value="Fri May 10 2019 00:00:00 GMT+0800 (CST)"
      // Exclude the impact of timezone
      scopedSlots: {
        content: '<div></div>',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('prop: editable', () => {
    wrapper = shallowMount(DatePicker, {
      propsData: {
        value: new Date(2019, 4, 10),
        defaultValue: new Date(2019, 4, 10),
        editable: false,
      },
      scopedSlots: {
        content: '<div></div>',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('prop: attrs of input', () => {
    wrapper = shallowMount(DatePicker, {
      propsData: {
        defaultValue: new Date(2019, 4, 10),
        placeholder: 'test placeholder',
        inputClass: 'test',
        inputAttr: {
          type: 'number',
          name: 'test',
        },
      },
      scopedSlots: {
        content: '<div></div>',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('prop: format', () => {
    wrapper = shallowMount(DatePicker, {
      propsData: {
        format: 'YYYY/MM/DD',
        value: new Date(2019, 9, 10),
      },
    });
    const input = wrapper.find('input').element;
    expect(input.value).toBe('2019/10/10');
  });

  it('prop: formatter', () => {
    wrapper = mount(DatePicker, {
      propsData: {
        valueType: 'format',
        value: '13/10/2019',
        open: true,
        type: 'week',
        formatter: {
          stringify(date) {
            return format(date, 'dd/MM/yyyy');
          },
          parse(value) {
            return parse(value, 'dd/MM/yyyy', new Date());
          },
          getWeek(date) {
            return date.getDate();
          },
        },
      },
    });
    const input = wrapper.find('input').element;
    expect(input.value).toBe('13/10/2019');
    const tableDate = wrapper.find('.mx-table-date');
    expect(tableDate.element).toMatchSnapshot();
  });

  it('prop: valueType', async () => {
    const value = new Date(2019, 9, 20);
    const emitValue = new Date(2019, 9, 22);
    wrapper = shallowMount(DatePicker, {
      propsData: {
        value,
        format: 'YYYY/MM/DD',
      },
    });
    const { vm } = wrapper;
    expect(vm.text).toBe('2019/10/20');
    vm.emitValue(emitValue);
    await wrapper.setProps({ valueType: 'format', value: '2019/10/20' });
    expect(vm.text).toBe('2019/10/20');
    vm.emitValue(emitValue);
    await wrapper.setProps({ valueType: 'timestamp', value: value.getTime() });
    expect(vm.text).toBe('2019/10/20');
    vm.emitValue(emitValue);
    await wrapper.setProps({ valueType: 'DD/MM/YYYY', value: '20/10/2019' });
    expect(vm.text).toBe('2019/10/20');
    vm.emitValue(emitValue);
    const emitted = wrapper.emitted();
    expect(emitted.input).toEqual([
      [emitValue],
      ['2019/10/22'],
      [emitValue.getTime()],
      ['22/10/2019'],
    ]);
  });

  it('prop: shortcut', async () => {
    const date = new Date(2019, 4, 10);
    wrapper = shallowMount(DatePicker, {
      propsData: {
        open: true,
        valueType: 'YYYY/MM/DD',
        range: false,
        shortcuts: [
          {
            text: 'Today',
            onClick() {
              return date;
            },
          },
        ],
      },
    });
    const btn = wrapper.find('.mx-btn-shortcut');
    await btn.trigger('click');
    const emitted = wrapper.emitted();
    expect(emitted.input).toEqual([['2019/05/10']]);
    await wrapper.setProps({
      range: true,
      shortcuts: [
        {
          text: 'range',
          onClick() {
            return [date, date];
          },
        },
      ],
    });
    btn.trigger('click');
    expect(emitted.input[1]).toEqual([['2019/05/10', '2019/05/10']]);
  });

  it('prop: popupClass', () => {
    wrapper = mount(DatePicker, {
      propsData: {
        open: true,
        popupClass: 'test',
      },
    });
    const popup = wrapper.findComponent(Popup);
    expect(popup.classes()).toContain('test');
  });

  it('prop: popupStyle', () => {
    wrapper = mount(DatePicker, {
      propsData: {
        open: true,
        popupStyle: {
          color: 'red',
        },
      },
    });
    const popup = wrapper.findComponent(Popup);
    expect(popup.element.style.color).toBe('red');
  });

  it('prop: confirm', async () => {
    wrapper = shallowMount(DatePicker, {
      propsData: {
        confirm: true,
      },
    });
    const { vm } = wrapper;
    await wrapper.find('input').trigger('focus');
    const btn = wrapper.find('.mx-datepicker-btn-confirm');
    expect(btn.exists()).toBe(true);
    // click the date expect popup don't close
    vm.handleSelectDate(new Date(2018, 5, 5));
    expect(wrapper.emitted().input).toBeUndefined();
    await btn.trigger('click');
    expect(wrapper.emitted().input[0][0]).toEqual(new Date(2018, 5, 5));
    expect(vm.popupVisible).toBe(false);
  });

  it('prop: confirmText', () => {
    wrapper = shallowMount(DatePicker, {
      propsData: {
        open: true,
        confirm: true,
        confirmText: 'test',
      },
    });

    const btn = wrapper.find('.mx-datepicker-btn-confirm');
    expect(btn.text()).toBe('test');
  });

  it('prop: appendToBody', () => {
    wrapper = mount(DatePicker, {
      propsData: {
        open: true,
        appendToBody: true,
      },
    });

    const popup = wrapper.findComponent(Popup);
    expect(popup.element.parentNode).toBe(document.body);
  });

  it('feat: should emit clear event when click clear button', async () => {
    wrapper = shallowMount(DatePicker, {
      propsData: {
        value: new Date(2019, 10, 9),
      },
    });
    await wrapper.find('.mx-input-wrapper').trigger('mouseenter');
    const clearButton = wrapper.find('.mx-icon-clear');

    clearButton.trigger('click');
    const emitted = wrapper.emitted();
    expect(emitted.clear).toBeTruthy();
    expect(emitted.input[0][0]).toBe(null);
  });

  it('feat: should emit [null, null] when clear range', async () => {
    wrapper = shallowMount(DatePicker, {
      propsData: {
        range: true,
        value: [new Date(2019, 10, 9), new Date(2019, 11, 9)],
      },
    });
    await wrapper.find('.mx-input-wrapper').trigger('mouseenter');
    const clearButton = wrapper.find('.mx-icon-clear');
    clearButton.trigger('click');
    const emitted = wrapper.emitted();
    expect(emitted.input[0][0]).toEqual([null, null]);
  });

  it('feat: should close popup when click time in datetime mode', () => {
    wrapper = mount(DatePicker, {
      propsData: {
        type: 'datetime',
        timePickerOptions: {
          start: '00:00',
          step: '00:30',
          end: '23:30',
        },
        open: true,
        showTimePanel: true,
      },
    });
    const el = wrapper.find('.mx-time-option');
    el.trigger('click');
    expect(wrapper.emitted().close).toBeTruthy();
  });

  // present the button submit form
  it('the type of all buttons should be button', () => {
    wrapper = mount(DatePicker, {
      propsData: {
        open: true,
        showTimePanel: true,
      },
    });
    const els = wrapper.findAll('button');
    els.wrappers.forEach(v => {
      expect(v.element.type).toBe('button');
    });
  });

  it('should emit pick event on first click', () => {
    wrapper = mount(DatePicker, {
      propsData: {
        range: true,
        open: true,
        defaultValue: new Date(2019, 9, 1),
      },
    });
    const td = wrapper.find('.mx-table-date td');
    td.trigger('click');
    expect(wrapper.emitted().pick[0][0]).toEqual(new Date(2019, 8, 29));
  });

  it('Ignore whitespace around separator on manual range input', async () => {
    const rangeSeparator = ' ~ ';
    const text = '2020-02-12';
    wrapper = mount(DatePicker, {
      propsData: {
        range: true,
        rangeSeparator: ' ~ ',
        valueType: 'format',
      },
    });
    const input = wrapper.find('input');

    await input.setValue(`${text} ${rangeSeparator} ${text}`);
    await input.trigger('change');
    await input.setValue(`${text}${rangeSeparator.trim()}${text}`);
    await input.trigger('change');
    await wrapper.setProps({ rangeSeparator: ' - ' });
    await input.setValue(`${text} - ${text}`);
    await input.trigger('change');
    expect(wrapper.emitted().input).toEqual([[[text, text]], [[text, text]], [[text, text]]]);
  });

  it('prop: multiple', () => {
    const value = [new Date(2020, 5, 6), new Date(2020, 6, 7)];
    wrapper = mount(DatePicker, {
      propsData: {
        multiple: true,
        open: true,
        value,
      },
    });
    wrapper.find('.mx-date-row .active').trigger('click');
    expect(wrapper.emitted().input[0][0]).toEqual(value.slice(0, 1));
    wrapper.find('[title="2020-07-15"]').trigger('click');
    expect(wrapper.emitted().input[1][0]).toEqual(value.concat(new Date(2020, 6, 15)));
  });

  it('prop: invalid multiple', () => {
    wrapper = shallowMount(DatePicker, {
      propsData: {
        multiple: true,
        range: true,
      },
    });
    const { vm } = wrapper;
    expect(vm.validMultipleType).toBe(false);
    wrapper.setProps({
      range: false,
      type: 'datetime',
    });
    expect(vm.validMultipleType).toBe(false);
  });

  it('If the value entered manually is in the disabled range should be invalid', () => {
    const someday = new Date(2020, 6, 1);
    wrapper = shallowMount(DatePicker, {
      format: 'YYYY-MM-DD',
      propsData: {
        disabledDate: date => {
          return date < someday;
        },
      },
    });
    const textInput = wrapper.find('input');
    textInput.setValue('2020-08-01');
    textInput.trigger('change');
    expect(wrapper.emitted().input[0][0]).toEqual(new Date(2020, 7, 1));
    textInput.setValue('2020-05-01');
    textInput.trigger('change');
    expect(wrapper.emitted().input[1]).toBe(undefined);
    expect(wrapper.emitted()['input-error'][0][0]).toBe('2020-05-01');
  });

  it('panel-change event should be emitd', async () => {
    wrapper = mount(DatePicker, {
      propsData: {
        open: true,
      },
    });
    const yearBtn = wrapper.find('.mx-btn-current-year');
    await yearBtn.trigger('click');
    expect(wrapper.emitted()['panel-change'][0]).toEqual(['year', 'date']);
  });
});
