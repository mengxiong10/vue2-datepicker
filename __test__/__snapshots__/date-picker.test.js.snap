// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker prop: attrs of input 1`] = `
<div
  class="mx-datepicker"
>
  <div
    class="mx-input-wrapper"
  >
    <input
      autocomplete="off"
      class="test"
      name="test"
      placeholder="test placeholder"
      type="number"
    />
    <i
      class="mx-icon-calendar"
    >
      <anonymous-stub />
    </i>
  </div>
  <popup-stub
    appendtobody="true"
  >
    <div
      class="mx-datepicker-content"
    >
      <div
        class="mx-datepicker-body"
      >
        <div />
      </div>
    </div>
  </popup-stub>
</div>
`;

exports[`DatePicker prop: clearable 1`] = `
<div
  class="mx-datepicker"
>
  <div
    class="mx-input-wrapper"
  >
    <input
      autocomplete="off"
      class="mx-input"
      name="date"
      placeholder=""
      type="text"
    />
    <i
      class="mx-icon-calendar"
    >
      <anonymous-stub />
    </i>
  </div>
  <popup-stub
    appendtobody="true"
  >
    <div
      class="mx-datepicker-content"
    >
      <div
        class="mx-datepicker-body"
      >
        <div />
      </div>
    </div>
  </popup-stub>
</div>
`;

exports[`DatePicker prop: editable 1`] = `
<div
  class="mx-datepicker"
>
  <div
    class="mx-input-wrapper"
  >
    <input
      autocomplete="off"
      class="mx-input"
      name="date"
      placeholder=""
      readonly="readonly"
      type="text"
    />
    <i
      class="mx-icon-clear"
    >
      <anonymous-stub />
    </i>
    <i
      class="mx-icon-calendar"
    >
      <anonymous-stub />
    </i>
  </div>
  <popup-stub
    appendtobody="true"
  >
    <div
      class="mx-datepicker-content"
    >
      <div
        class="mx-datepicker-body"
      >
        <div />
      </div>
    </div>
  </popup-stub>
</div>
`;

exports[`DatePicker prop: required 1`] = `
<div
  class="mx-datepicker"
>
  <div
    class="mx-input-wrapper"
  >
    <input
      autocomplete="off"
      class="mx-input"
      name="date"
      placeholder=""
      required="required"
      type="text"
    />
    <i
      class="mx-icon-clear"
    >
      <anonymous-stub />
    </i>
    <i
      class="mx-icon-calendar"
    >
      <anonymous-stub />
    </i>
  </div>
  <popup-stub
    appendtobody="true"
  >
    <div
      class="mx-datepicker-content"
    >
      <div
        class="mx-datepicker-body"
      >
        <div />
      </div>
    </div>
  </popup-stub>
</div>
`;

exports[`DatePicker prop: formatter 1`] = `
<table
  class="mx-table mx-table-date"
>
  <thead>
    <tr>
      <th
        class="mx-week-number-header"
      />
       
      <th>
        Su
      </th>
      <th>
        Mo
      </th>
      <th>
        Tu
      </th>
      <th>
        We
      </th>
      <th>
        Th
      </th>
      <th>
        Fr
      </th>
      <th>
        Sa
      </th>
    </tr>
  </thead>
   
  <tbody>
    <tr
      class="mx-date-row"
    >
      <td
        class="mx-week-number"
        data-row-col="0,0"
      >
        
            29
          
      </td>
       
      <td
        class="cell not-current-month"
        data-row-col="0,0"
        title="2019-09-29"
      >
        <div>
          29
        </div>
      </td>
      <td
        class="cell not-current-month"
        data-row-col="0,1"
        title="2019-09-30"
      >
        <div>
          30
        </div>
      </td>
      <td
        class="cell"
        data-row-col="0,2"
        title="2019-10-01"
      >
        <div>
          1
        </div>
      </td>
      <td
        class="cell"
        data-row-col="0,3"
        title="2019-10-02"
      >
        <div>
          2
        </div>
      </td>
      <td
        class="cell"
        data-row-col="0,4"
        title="2019-10-03"
      >
        <div>
          3
        </div>
      </td>
      <td
        class="cell"
        data-row-col="0,5"
        title="2019-10-04"
      >
        <div>
          4
        </div>
      </td>
      <td
        class="cell"
        data-row-col="0,6"
        title="2019-10-05"
      >
        <div>
          5
        </div>
      </td>
    </tr>
    <tr
      class="mx-date-row"
    >
      <td
        class="mx-week-number"
        data-row-col="1,0"
      >
        
            6
          
      </td>
       
      <td
        class="cell"
        data-row-col="1,0"
        title="2019-10-06"
      >
        <div>
          6
        </div>
      </td>
      <td
        class="cell"
        data-row-col="1,1"
        title="2019-10-07"
      >
        <div>
          7
        </div>
      </td>
      <td
        class="cell"
        data-row-col="1,2"
        title="2019-10-08"
      >
        <div>
          8
        </div>
      </td>
      <td
        class="cell"
        data-row-col="1,3"
        title="2019-10-09"
      >
        <div>
          9
        </div>
      </td>
      <td
        class="cell"
        data-row-col="1,4"
        title="2019-10-10"
      >
        <div>
          10
        </div>
      </td>
      <td
        class="cell"
        data-row-col="1,5"
        title="2019-10-11"
      >
        <div>
          11
        </div>
      </td>
      <td
        class="cell"
        data-row-col="1,6"
        title="2019-10-12"
      >
        <div>
          12
        </div>
      </td>
    </tr>
    <tr
      class="mx-date-row mx-active-week"
    >
      <td
        class="mx-week-number"
        data-row-col="2,0"
      >
        
            13
          
      </td>
       
      <td
        class="cell active"
        data-row-col="2,0"
        title="2019-10-13"
      >
        <div>
          13
        </div>
      </td>
      <td
        class="cell"
        data-row-col="2,1"
        title="2019-10-14"
      >
        <div>
          14
        </div>
      </td>
      <td
        class="cell"
        data-row-col="2,2"
        title="2019-10-15"
      >
        <div>
          15
        </div>
      </td>
      <td
        class="cell"
        data-row-col="2,3"
        title="2019-10-16"
      >
        <div>
          16
        </div>
      </td>
      <td
        class="cell"
        data-row-col="2,4"
        title="2019-10-17"
      >
        <div>
          17
        </div>
      </td>
      <td
        class="cell"
        data-row-col="2,5"
        title="2019-10-18"
      >
        <div>
          18
        </div>
      </td>
      <td
        class="cell"
        data-row-col="2,6"
        title="2019-10-19"
      >
        <div>
          19
        </div>
      </td>
    </tr>
    <tr
      class="mx-date-row"
    >
      <td
        class="mx-week-number"
        data-row-col="3,0"
      >
        
            20
          
      </td>
       
      <td
        class="cell"
        data-row-col="3,0"
        title="2019-10-20"
      >
        <div>
          20
        </div>
      </td>
      <td
        class="cell"
        data-row-col="3,1"
        title="2019-10-21"
      >
        <div>
          21
        </div>
      </td>
      <td
        class="cell"
        data-row-col="3,2"
        title="2019-10-22"
      >
        <div>
          22
        </div>
      </td>
      <td
        class="cell"
        data-row-col="3,3"
        title="2019-10-23"
      >
        <div>
          23
        </div>
      </td>
      <td
        class="cell"
        data-row-col="3,4"
        title="2019-10-24"
      >
        <div>
          24
        </div>
      </td>
      <td
        class="cell"
        data-row-col="3,5"
        title="2019-10-25"
      >
        <div>
          25
        </div>
      </td>
      <td
        class="cell"
        data-row-col="3,6"
        title="2019-10-26"
      >
        <div>
          26
        </div>
      </td>
    </tr>
    <tr
      class="mx-date-row"
    >
      <td
        class="mx-week-number"
        data-row-col="4,0"
      >
        
            27
          
      </td>
       
      <td
        class="cell"
        data-row-col="4,0"
        title="2019-10-27"
      >
        <div>
          27
        </div>
      </td>
      <td
        class="cell"
        data-row-col="4,1"
        title="2019-10-28"
      >
        <div>
          28
        </div>
      </td>
      <td
        class="cell"
        data-row-col="4,2"
        title="2019-10-29"
      >
        <div>
          29
        </div>
      </td>
      <td
        class="cell"
        data-row-col="4,3"
        title="2019-10-30"
      >
        <div>
          30
        </div>
      </td>
      <td
        class="cell"
        data-row-col="4,4"
        title="2019-10-31"
      >
        <div>
          31
        </div>
      </td>
      <td
        class="cell not-current-month"
        data-row-col="4,5"
        title="2019-11-01"
      >
        <div>
          1
        </div>
      </td>
      <td
        class="cell not-current-month"
        data-row-col="4,6"
        title="2019-11-02"
      >
        <div>
          2
        </div>
      </td>
    </tr>
    <tr
      class="mx-date-row"
    >
      <td
        class="mx-week-number"
        data-row-col="5,0"
      >
        
            3
          
      </td>
       
      <td
        class="cell not-current-month"
        data-row-col="5,0"
        title="2019-11-03"
      >
        <div>
          3
        </div>
      </td>
      <td
        class="cell not-current-month"
        data-row-col="5,1"
        title="2019-11-04"
      >
        <div>
          4
        </div>
      </td>
      <td
        class="cell not-current-month"
        data-row-col="5,2"
        title="2019-11-05"
      >
        <div>
          5
        </div>
      </td>
      <td
        class="cell not-current-month"
        data-row-col="5,3"
        title="2019-11-06"
      >
        <div>
          6
        </div>
      </td>
      <td
        class="cell not-current-month"
        data-row-col="5,4"
        title="2019-11-07"
      >
        <div>
          7
        </div>
      </td>
      <td
        class="cell not-current-month"
        data-row-col="5,5"
        title="2019-11-08"
      >
        <div>
          8
        </div>
      </td>
      <td
        class="cell not-current-month"
        data-row-col="5,6"
        title="2019-11-09"
      >
        <div>
          9
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
